---
title: inferrnal
author: Brendan Furneaux
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{inferrnal}
    \usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

R Interface to Call Programs from Infernal RNA Covariance Model Package

Covariance Models (CM) are stochastic models of RNA sequence and secondary
structure.
[Infernal](http://eddylab.org/infernal/) (INFERence of RNA ALignment) is a
software package with various command-line tools related to CMs.
`inferrnal` (with two "`r`"s) is a lightweight R interface which calls the
Infernal tools and imports the results to R.
It is developed independently from Infernal, and Infernal must be installed in
order for it to function.
Note that Infernal does not work on Windows.

# Examples

So far only two of the tools are implemented.

## cmsearch

In order to search, we need a CM.
[Rfam](https://rfam.xfam.org/) has a wide variety.
For this example, we will use the eukaryotic 5.8S rRNA, with the Rfam ID
[`RF00002`](https://rfam.xfam.org/family/RF00002).
The CM is [available from Rfam](https://rfam.xfam.org/family/RF00002/cm), but
it is also included as example data in `inferrnal`.


```{r cm}
library(inferrnal)
cm <- cm_5_8S()
```

We also need some sequences to search.
The sample data is from a soil metabarcoding study focused on fungi.
The targeted region includes 5.8S as well as some of the surrounding rDNA
regions.

```{r reads}
sampfasta <- sample_rRNA_fasta()
```

Use `cmsearch()` to locate the 5.8S RNA in each sequence.

```{r cmsearch}
library(inferrnal)
cmsearch(cm = cm, seq = sampfasta, cpu = 1)
```

Instead of passing a file name, you can also supply a `DNAStringSet` or
`RNAStringSet` object from the `Biostrings` package.

```{r cmsearch-seq}
sampseqs <- Biostrings::readDNAStringSet(sampfasta)
cmsearch(cm = cm, seq = sampseqs, cpu = 1)
```

`cmsearch`, by default, returns a table with information about each hit.
However, it can optionally also output an alignment of the hits in Stockholm
format.

```{r cmsearch-aln}
alnfile <- tempfile("alignment-", fileext = ".stk")
cmsearch(cm = cm, seq = sampseqs, alignment = alnfile)
```

`inferrnal` includes a simple parser for Stockholm alignments, which also
imports column annotations.
```{r read_msa}
msa <- read_stockholm_msa(alnfile)
```

`read_stockholm_msa` returns an object of class
`StockholmRNAMultipleSequenceAlignment`.

```{r msa_align}
msa
```

Other useful information which Infernal outputs include the consensus secondary
structure and the reference annotation, both of which are defined by the CM.
These are given as column ("GC") annotations named "SS_cons" and "RF",
respectively, in the Stockholm alignment file.

## cmalign

If you have sequences which have already been trimmed to contain only the RNA
defined by the CM (possibly truncated, but not extended), then you can align
them to the CM using `cmalign`.
This is much faster than `cmsearch`.
This example uses the results of `cmsearch` from the previous section, after
removing gaps.

```{r}
unaln <- sample_rRNA_5_8S()
unaln_seq <- Biostrings::readRNAStringSet(unaln)
unaln_seq
aln <- cmalign(cm, unaln_seq, cpu = 1)
aln
```

## cmbuild

`cmbuild`, included in `inferrnal` since version 0.99.5, is used to create new
CMs from annotated multiple sequence alignments.
To illustrate the process, we use the seed alignment for the 5.8S rRNA CM from
RFAM.
It is included as a sample file in `inferrnal`.

```{r}
new_cm <- file.path(tempdir(), "5_8S.cm")
cmbuild(new_cm, msafile = stk_5_8S(), force = TRUE, quiet = FALSE)
```

This CM is not calibrated, so it cannot be used for `cmsearch`,
but it can be used in `cmalign`.

```{r}
aln2 <- cmalign(new_cm, unaln_seq, cpu = 1)
```

The resulting alignment is the same as using the CM which was generated by RFAM,
because they are based on the same seed alignment, and used the same (default)
options for `cmbuild`.

```{r}
all.equal(aln, aln2)
```
